### ISSUE TITLE
Shader compiler prevents Bevy from compiling on Raspberry Pi 4

### ISSUE THREAD

**Issue by Type1J** [bevyengine/bevy#2881]
## Bevy version
0.5.0

## Operating system & version

Manjaro Gnome 21.1.4 on Raspberry Pi 4B+ with 8GB RAM

## What you did

Created a project with Bevy as a dependency:
```cargo
[package]
name = "test_bevy"
version = "0.1.0"
edition = "2018"

[dependencies]
bevy = "0.5.0"
```
Then I ran `cargo build` in the new project.

## What you expected to happen

Bevy should compile on Raspberry Pi 4.

## What actually happened

Then `cargo build` gives:
```
...
error: failed to run custom build command for `bevy-glsl-to-spirv v0.2.1`

Caused by:
  process didn't exit successfully: `/home/jay/Projects/led-console-service/target/debug/build/bevy-glsl-to-spirv-b0e467b111f4a87e/build-script-build` (exit status: 101)
  --- stderr
  thread 'main' panicked at 'Missing target support aarch64-unknown-linux-gnu', /home/jay/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy-glsl-to-spirv-0.2.1/build/build.rs:24:9
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed
```

## Additional information

There seems to be iOS and Android support using a different internal dependency. I'm not sure if `bevy-glsl-to-spirv` is the correct dependency, but it might be, since this is GNU/Linux, not Android/Linux. I'll post additional info when I attempt to compile Bevy from the Git repo at the 0.5.0 release hash.

**Comment by bjorn3** [comment]
The next bevy release will likely use the pure rust crate naga instead of bevy-glsl-to-spriv which uses C++.

**Comment by Type1J** [comment]
@bjorn3 Is that what's in the `main` branch, now, or is that in a feature branch that I could test?

**Comment by alice-i-cecile** [comment]
#2556 is likely the branch you'll need to test; it updates to a newer WGPU version in the hopes of getting OpenGL support working.

**Comment by Type1J** [comment]
@alice-i-cecile The branch that you suggested didn't compile on Raspberry Pi, but maybe you know why from the error messages.

It seems that the correct backend is not selected, so it tries to compile all of them... I think; I'm really not sure. Here's the error messages:
```
   Compiling wgpu-hal v0.10.7
error: Metal API enabled on non-Apple OS. If your project is not using resolver="2" in Cargo.toml, it should.
  --> /home/jay/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-hal-0.10.7/src/lib.rs:45:1
   |
45 | compile_error!("Metal API enabled on non-Apple OS. If your project is not using resolver=\"2\" in Cargo.toml, it should.");
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: DX12 API enabled on non-Windows OS. If your project is not using resolver="2" in Cargo.toml, it should.
  --> /home/jay/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-hal-0.10.7/src/lib.rs:47:1
   |
47 | compile_error!("DX12 API enabled on non-Windows OS. If your project is not using resolver=\"2\" in Cargo.toml, it should.");
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0432]: unresolved import `super::dx12`
  --> /home/jay/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-hal-0.10.7/src/lib.rs:62:20
   |
62 |     pub use super::dx12::Api as Dx12;
   |                    ^^^^ could not find `dx12` in the crate root

error[E0432]: unresolved import `super::metal`
  --> /home/jay/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-hal-0.10.7/src/lib.rs:67:20
   |
67 |     pub use super::metal::Api as Metal;
   |                    ^^^^^ could not find `metal` in the crate root

For more information about this error, try `rustc --explain E0432`.
error: could not compile `wgpu-hal` due to 4 previous errors
warning: build failed, waiting for other jobs to finish...
error: build failed
```
Is there a feature that I need to enable? I'm assuming so, due to the `If your project is not using resolver="2" in Cargo.toml, it should.` message, but I'm not sure what that means.

**Comment by Type1J** [comment]
I just read `https://doc.rust-lang.org/cargo/reference/resolver.html` and I'll try with `resolver = "2"`.

**Comment by Type1J** [comment]
Ok, new error messages:
```
   Compiling bevy_render v0.5.0 (/home/jay/Projects/bevy/update/crates/bevy_render)
   Compiling bevy_render2 v0.5.0 (/home/jay/Projects/bevy/update/pipelined/bevy_render2)
   Compiling bevy_scene v0.5.0 (/home/jay/Projects/bevy/update/crates/bevy_scene)
error[E0599]: no associated item named `VERTEX` found for struct `shader::shader::ShaderStages` in the current scope
   --> /home/jay/Projects/bevy/update/crates/bevy_render/src/shader/shader.rs:109:27
    |
109 |             ShaderStages::VERTEX => shaderc::ShaderKind::Vertex,
    |                           ^^^^^^ associated item not found in `shader::shader::ShaderStages`
...
248 | pub struct ShaderStages {
    | ----------------------- associated item `VERTEX` not found for this
    |
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following traits define an item `VERTEX`, perhaps you need to implement one of them:
            candidate #1: `<binding::BindingShaderStage as Debug>::fmt::__BitFlags`
            candidate #2: `binding::BindingShaderStage::all::__BitFlags`
            candidate #3: `<renderer::render_resource::buffer::BufferUsage as Debug>::fmt::__BitFlags`
            candidate #4: `renderer::render_resource::buffer::BufferUsage::all::__BitFlags`

error[E0599]: no associated item named `FRAGMENT` found for struct `shader::shader::ShaderStages` in the current scope
   --> /home/jay/Projects/bevy/update/crates/bevy_render/src/shader/shader.rs:110:27
    |
110 |             ShaderStages::FRAGMENT => shaderc::ShaderKind::Fragment,
    |                           ^^^^^^^^ associated item not found in `shader::shader::ShaderStages`
...
248 | pub struct ShaderStages {
    | ----------------------- associated item `FRAGMENT` not found for this
    |
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following traits define an item `FRAGMENT`, perhaps you need to implement one of them:
            candidate #1: `<binding::BindingShaderStage as Debug>::fmt::__BitFlags`
            candidate #2: `binding::BindingShaderStage::all::__BitFlags`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bevy_render` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by Type1J** [comment]
Ok, so after adding aarch64/linux/gnu everywhere I saw android and x86_64 linux together in the `cfg` blocks and Cargo conditional compilation, I got back to this:
```
   Compiling bevy-glsl-to-spirv v0.2.1
error: failed to run custom build command for `bevy-glsl-to-spirv v0.2.1`

Caused by:
  process didn't exit successfully: `/home/jay/Projects/led-console-service/target/debug/build/bevy-glsl-to-spirv-b0e467b111f4a87e/build-script-build` (exit status: 101)
  --- stderr
  thread 'main' panicked at 'Missing target support aarch64-unknown-linux-gnu', /home/jay/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy-glsl-to-spirv-0.2.1/build/build.rs:24:9
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by Type1J** [comment]
It did compile once I got the `bevy-glsl-to-spirv` repo and added `aarch64-linux-unknown-gnu` to the `build.rs` file. I'm not sure if that was the intention of using the alternative branch, though. Also, I haven't used any graphics functions, yet. I'll try those out, and if they work, I'll make a PR to `bevy-glsl-to-spirv`, even though it seems to be going away soon, and to bevy (there are still a few places where aarch64 had to be added).

**Comment by Type1J** [comment]
Ok, I'm stuck again. Projects using Bevy have this link error:
```
  = note: /usr/bin/ld: cannot find -lglslang.glsltospirv
          /usr/bin/ld: cannot find -lHLSL.glsltospirv
          /usr/bin/ld: cannot find -lOGLCompiler.glsltospirv
          /usr/bin/ld: cannot find -lOSDependent.glsltospirv
          /usr/bin/ld: cannot find -lSPIRV.glsltospirv
          /usr/bin/ld: cannot find -lSPVRemapper.glsltospirv
          /usr/bin/ld: cannot find -lSPIRV-Tools-opt.glsltospirv
          /usr/bin/ld: cannot find -lSPIRV-Tools.glsltospirv
          collect2: error: ld returned 1 exit status
 ```
I'm assuming that this is due to to the changes that I made still using `bevy-glsl-to-spirv` instead of `naga`. I'll attempt to see where `naga` is used in this branch, but if anybody knows where features/cfg/etc. are needed, please tell me.

**Comment by cart** [comment]
The issue is that currently `pipelined-rendering` has both the old renderer plugin and the new renderer plugin. Prior to the next bevy release, we will remove the old plugin entirely. But until that happens the old render plugin would need to be removed manually. Also note that by default, the old plugin is used. Some of the examples use the new plugin, which they opt in to by using PipelinedDefaultPlugins instead of DefaultPlugins.

**Comment by Type1J** [comment]
I saw. It's been interesting going through the code of Bevy, but I'll wait until the release for the Raspberry Pi 4 app that I was making.

Is the next release happening soon?

**Comment by alice-i-cecile** [comment]
> Is the next release happening soon?

For some definition of soon! We're currently wrapping up the rendering rework: see #2535 for progress and tasks to be done. Following that there'll be a bit of planning and tying up loose ends (merging PRs that are ready-to-go, making release notes) before we release.

**Comment by Type1J** [comment]
For now, I'm using some of the internal crates directly. I really only needed the timer, ECS, and a system that drives some ws2812 LEDs based on some data that I get from the network. There is a touchscreen on it, too, and I was hoping to render what the LED array showed, but I can get that working on an x86_64 Linux machine and recompile it later.

By the way, Bevy is a great engine! I'm going to be using it in a lot of my projects.

**Comment by Type1J** [comment]
I'll close this issue when I can render graphics on the Raspberry Pi 4. It sounds like that will be 0.6.0 at the soonest. Bu the way, if it works on Raspberry Pi, it will probably also work on Pinebook and Pinebook Pro (Rock64).

**Comment by Type1J** [comment]
@inodentry It did select shaderc, I believe. When it attempted to build, (copying from above) I got these errors:
```
   Compiling bevy_render v0.5.0 (/home/jay/Projects/bevy/update/crates/bevy_render)
   Compiling bevy_render2 v0.5.0 (/home/jay/Projects/bevy/update/pipelined/bevy_render2)
   Compiling bevy_scene v0.5.0 (/home/jay/Projects/bevy/update/crates/bevy_scene)
error[E0599]: no associated item named `VERTEX` found for struct `shader::shader::ShaderStages` in the current scope
   --> /home/jay/Projects/bevy/update/crates/bevy_render/src/shader/shader.rs:109:27
    |
109 |             ShaderStages::VERTEX => shaderc::ShaderKind::Vertex,
    |                           ^^^^^^ associated item not found in `shader::shader::ShaderStages`
...
248 | pub struct ShaderStages {
    | ----------------------- associated item `VERTEX` not found for this
    |
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following traits define an item `VERTEX`, perhaps you need to implement one of them:
            candidate #1: `<binding::BindingShaderStage as Debug>::fmt::__BitFlags`
            candidate #2: `binding::BindingShaderStage::all::__BitFlags`
            candidate #3: `<renderer::render_resource::buffer::BufferUsage as Debug>::fmt::__BitFlags`
            candidate #4: `renderer::render_resource::buffer::BufferUsage::all::__BitFlags`

error[E0599]: no associated item named `FRAGMENT` found for struct `shader::shader::ShaderStages` in the current scope
   --> /home/jay/Projects/bevy/update/crates/bevy_render/src/shader/shader.rs:110:27
    |
110 |             ShaderStages::FRAGMENT => shaderc::ShaderKind::Fragment,
    |                           ^^^^^^^^ associated item not found in `shader::shader::ShaderStages`
...
248 | pub struct ShaderStages {
    | ----------------------- associated item `FRAGMENT` not found for this
    |
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following traits define an item `FRAGMENT`, perhaps you need to implement one of them:
            candidate #1: `<binding::BindingShaderStage as Debug>::fmt::__BitFlags`
            candidate #2: `binding::BindingShaderStage::all::__BitFlags`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bevy_render` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: build failed
```

I tried changing some of the struct field names, but whatever I did seemed to have the same complaint, so I started grasping at straws, and changed the conditional compile guards to enable bevy-glsl-to-spirv, but it unsurprisingly had the same errors as the published crate.

**Comment by nickelser** [comment]
Any progress on getting bevy to compile on rpi4? Was about to try this myself but wondering if it's worth it.

**Comment by Type1J** [comment]
It does compile if you don't use the a renderer. I have used it on the Pi4 in a way that doesn't use graphics.

I'm pretty confident that it will be easier once the `pipelined-rendering` branch removes the old renderer. I'm waiting on that change before I try anything else.

**Comment by cart** [comment]
`pipelined-rendering` has been merged into `main`. The old renderer is still compiled by default, but I just merged a change to the iOS example that shows how to opt out of compiling the old renderer: https://github.com/bevyengine/bevy/blob/main/examples/ios/Cargo.toml. You should be able to test this now (combined with something like the `3d_scene_pipelined` example)

**Comment by mockersf** [comment]
> I just merged a change to the iOS example that shows how to opt out of compiling the old renderer

Not yet: #3232

**Comment by nickelser** [comment]
> I have used it on the Pi4 in a way that doesn't use graphics.

I am curious what use case this is! But thank you both for the info -- I'll give it a shot this week with the new pipelined renderer.

**Comment by Type1J** [comment]
> I am curious what use case this is!
The Raspberry Pi is being used to drive WS2812 Addressable LED strips for visual output.

**Comment by nickelser** [comment]
> The Raspberry Pi is being used to drive WS2812 Addressable LED strips for visual output.

Oh I missed your message earlier in the thread, sorry. That's very cool.

With `main` I am able to get `breakout` to compile and run; it looks _extremely_ wonky though (see screenshot). Should I create a separate issue for this as I believe that this issue is resolved?

![IMG_6435](https://user-images.githubusercontent.com/367553/144718351-108a3202-f1cc-46c7-9560-cf5e60081c2e.PNG)

There is also a warning about v3dv, which is perhaps the issue.

```
pi@raspberrypi:~/dev/bevy $ cargo run --example breakout
    Finished dev [unoptimized + debuginfo] target(s) in 1.05s
     Running `target/debug/examples/breakout`
2021-12-04T00:25:50.434921Z  WARN winit::platform_impl::platform::x11::util::randr: XRandR reported that the display's 0mm in size, which is certifiably insane    
2021-12-04T00:25:50.435610Z  INFO winit::platform_impl::platform::x11::window: Guessed window scale factor: 1    
WARNING: v3dv is neither a complete nor a conformant Vulkan implementation. Testing use only.
2021-12-04T00:25:51.585211Z  INFO naga::front::spv: Generated by 851978 version 10000    
2021-12-04T00:25:51.591875Z  INFO naga::front::spv: Patching...    
2021-12-04T00:25:51.594482Z  INFO naga::front::spv: Generated by 851978 version 10000    
2021-12-04T00:25:51.596458Z  INFO naga::front::spv: Patching...    
2021-12-04T00:25:53.009906Z  INFO naga::front::spv: Generated by 851978 version 10000    
2021-12-04T00:25:53.018755Z  INFO naga::front::spv: Patching...    
2021-12-04T00:25:53.020855Z  INFO naga::front::spv: Generated by 851978 version 10000    
2021-12-04T00:25:53.022608Z  INFO naga::front::spv: Patching... 
```

**Comment by NiklasEi** [comment]
I would say this issue is resolved as soon as Bevy does no longer compile the old renderer by default. Which I think is still the case?

@nickelser Feel free to open a new issue about the rendering problem though :+1:

**Comment by mockersf** [comment]
> `XRandR reported that the display's 0mm in size, which is certifiably insane`

I like that log!

> With main I am able to get breakout to compile and run

Did you change the code example, or was it using the old renderer?

**Comment by nickelser** [comment]
> > `XRandR reported that the display's 0mm in size, which is certifiably insane`
> 
> 
> 
> I like that log!

Haha yes. I suspect that was due to me running it via SSH (with `DISPLAY` set to the attached screen).

> Did you change the code example, or was it using the old renderer?

This was with no changes on `main` so I think that means it works with the old renderer too. I will try the new render examples later.

**Comment by alice-i-cecile** [comment]
Really cool stuff! I'm going to close this out, but I'm very curious to see the new renderer results. If there are other graphical issues, please open up new issues for us :)
